components/admin_console/team_channel_settings/group/group_users/group_users.tsx

```diff
@@ -0,0 +1,393 @@
+
+import UserGridName from 'components/admin_console/user_grid/user_grid_name';
+import DataGrid, {Row, Column} from 'components/admin_console/data_grid/data_grid';
+import {FilterOptions} from 'components/admin_console/filter/filter';
+
+import GroupUsersRole from './group_users_role';
+import GroupUsersGroups from './group_users_groups';
+
+import './group_users.scss';
+
+const GROUP_MEMBERS_PAGE_SIZE = 10;
+
+export type Filters = {
+    roles?: string[];
+    channel_roles?: string[];
+    team_roles?: string[];
+}
+
+export type Memberships = RelationOneToOne<UserProfile, TeamMembership> | RelationOneToOne<UserProfile, ChannelMembership>;
+
+interface Props {
+    members: UserProfile[];
+    memberships: Memberships;
+    total: number;
+    searchTerm: string;
+    scope: 'team' | 'channel';
+    scopeId: string;
+    enableGuestAccounts: boolean;
+    filters: Filters;
+    actions: {
+        loadTeamMembersForProfilesList: (profiles: UserProfile[], teamId: string) => Promise<{
+            data: boolean;
+        }>;
+        loadChannelMembersForProfilesList: (profiles: UserProfile[], channelId: string) => Promise<{
+            data: boolean;
+        }>;
+        setModalSearchTerm: (term: string) => Promise<{
+            data: boolean;
+        }>;
+        setModalFilters: (filters: Filters) => Promise<{
+            data: boolean;
+        }>;
+    };
+}
+
+interface State {
+    page: number;
+    loading: boolean;
+}
+
+export default class GroupUsers extends React.PureComponent<Props, State> {
+    private searchTimeoutId: number;
+
+    public constructor(props: Props) {
+        super(props);
+        this.searchTimeoutId = 0;
+
+        this.state = {
+            page: 0,
+            loading: true,
+        };
+    }
+
+    componentDidUpdate(prevProps: Props) {
+        const searchTerm = this.props.searchTerm;
+        const filters = this.props.filters;
+        if (searchTerm !== prevProps.searchTerm || JSON.stringify(filters) !== JSON.stringify(prevProps.filters)) {
+            clearTimeout(this.searchTimeoutId);
+            this.setStateLoading(true, 0);
+
+            if (searchTerm === '' && Object.keys(filters).length === 0) {
+                this.setStateLoading(false);
+                return;
+            }
+
+            this.searchTimeoutId = window.setTimeout(
+                async () => {
+                    this.setStateLoading(false);
+                },
+                300,
+            );
+        }
+    }
+
+    async componentDidMount() {
+        const {members, total} = this.props;
+        const MEMBERSHIPS_TO_LOAD_COUNT = 100;
+        const promises = [];
+        let membershipsLoaded = 0;
+
+        // Pre-load all memberships since users are also already loaded into the state
+        while (membershipsLoaded < total) {
+            promises.push(this.loadMembersForProfilesList(members.slice(membershipsLoaded, membershipsLoaded + MEMBERSHIPS_TO_LOAD_COUNT)));
```

Comment:
Should the loading of channel members be done lazily as more pages are viewed? Loading them all at once seems like it could cause performance issues in larger channels/teams

Reply:
So my initial implementation had me loading the memberships lazily... 

_But_ theres no api endpoints to search channel members minus group members with the required filters too, to avoid having to do a bunch of server side work to add the support to those APIs I opted to just load all the memberships when this dialog is opened and do all the filtering locally, I think it shouldnt be too big of a deal since this modal will never really be _all_ of the channel or team members.. it just shows the members to be _removed_ when enabling group sync on a channel or team. I can test it out with a few thousand users later but it should be fine this component isnt worth the time investment at the moment I think.

Reply:
Is there perhaps an API to get channel members by their user IDs? That said, I agree that the performance issue won't be exactly huge, so it might not be something we need to address now.